packages/ses/src/compartment-shim.js

```diff
@@ -83,13 +89,143 @@ const assertModuleHooks = compartment => {
   }
 };
 
-/**
- * Compartment()
- * The Compartment constructor is a global. A host that wants to execute
- * code in a context bound to a new global creates a new compartment.
+/** SharedCompartmentConstructor
+ *
+ * All compartments share a prototype, though there are many compartment
+ * constructors.
+ * The compartment shim introduces a fully powered global Compartment constructor.
+ * The lockdown shim replaces that Compartment constructor with one whose descendants
+ * all share the new tamed and hardened environment.
+ * Within each Compartment, there is a unique Compartment constructor that
+ * closes over the parent.
+ *
+ * Before lockdown, we expect every constructor to have a prototype and for the
+ * prototype to refer back to the constructor.
+ * After lockdown, we expect the constructors from all compartments to share a
+ * prototype, but the prototype can no longer carry a reference back to any
+ * particular constructor.
+ * For example, every compartment has a unique and powerful Function
+ * constructor, each of these has the same prototype.
+ * Because the prototype is shared among compartments, the shared constructor
+ * on the prototype must be powerless.
+ *
+ * Following this pattern, the shared compartment constructor is also powerless
+ * and simply throws an exception.
+ */
+export const SharedCompartmentConstructor = function Compartment() {
+  throw new TypeError(
+    `Cannot call or construct compartment with shared constructor`,
+  );
+};
+
+SharedCompartmentConstructor.toString = () => {
+  return 'function Compartment() { [shim code] }';
+};
+
+/** CompartmentPrototype
+ * Every compartment constructor shares this prototype.
+ * As with other primordials, the
  */
-export class Compartment {
-  constructor(endowments = {}, modules = {}, options = {}) {
+const CompartmentPrototype = {
+  constructor: SharedCompartmentConstructor,
+
+  get globalThis() {
+    return privateFields.get(this).globalObject;
+  },
+
+  /**
+   * @param {string} source is a JavaScript program grammar construction.
+   * @param {{
+   *   transforms: Array<Transform>,
+   *   sloppyGlobalsMode: bool,
+   * }} options.
+   */
+  evaluate(source, options = {}) {
+    // Perform this check first to avoid unecessary sanitizing.
+    if (typeof source !== 'string') {
+      throw new TypeError('first argument of evaluate() must be a string');
+    }
+
+    // Extract options, and shallow-clone transforms.
+    const { transforms = [], sloppyGlobalsMode = false } = options;
+    const localTransforms = [...transforms];
+
+    const {
+      globalTransforms,
+      globalObject,
+      globalLexicals,
+    } = privateFields.get(this);
+    const realmRec = getCurrentRealmRec();
+
+    return performEval(realmRec, source, globalObject, globalLexicals, {
```

Comment:
Just reiterating that the only thing this module does with `realmRec` is pass it to `performEval`. Other than that, this module doesn't have too much magic authority. Better POLA would be for this module to ask for a `currentRealmPerformEval` or something that already encapsulates the needed `realmRec`.

Doesn't need to happen in this PR but we need a TODO or issue or something.

Reply:
https://github.com/Agoric/SES-shim/issues/385